components/pipsolar/pipsolar.h

```diff
@@ -0,0 +1,222 @@
+  PIPSOLAR_BINARY_SENSOR(fault_inverter_soft_failed, QPIWS, bool)
+  PIPSOLAR_BINARY_SENSOR(fault_self_test_failed, QPIWS, bool)
+  PIPSOLAR_BINARY_SENSOR(fault_op_dc_voltage_over, QPIWS, bool)
+  PIPSOLAR_BINARY_SENSOR(fault_battery_open, QPIWS, bool)
+  PIPSOLAR_BINARY_SENSOR(fault_current_sensor_failed, QPIWS, bool)
+  PIPSOLAR_BINARY_SENSOR(fault_battery_short, QPIWS, bool)
+  PIPSOLAR_BINARY_SENSOR(warning_power_limit, QPIWS, bool)
+  PIPSOLAR_BINARY_SENSOR(warning_pv_voltage_high, QPIWS, bool)
+  PIPSOLAR_BINARY_SENSOR(fault_mppt_overload, QPIWS, bool)
+  PIPSOLAR_BINARY_SENSOR(warning_mppt_overload, QPIWS, bool)
+  PIPSOLAR_BINARY_SENSOR(warning_battery_too_low_to_charge, QPIWS, bool)
+  PIPSOLAR_BINARY_SENSOR(fault_dc_dc_over_current, QPIWS, bool)
+  PIPSOLAR_BINARY_SENSOR(fault_code, QPIWS, int)
+  PIPSOLAR_BINARY_SENSOR(warnung_low_pv_energy, QPIWS, bool)
+  PIPSOLAR_BINARY_SENSOR(warning_high_ac_input_during_bus_soft_start, QPIWS, bool)
+  PIPSOLAR_BINARY_SENSOR(warning_battery_equalization, QPIWS, bool)
+
+  PIPSOLAR_TEXT_SENSOR(last_qpigs, QPIGS)
+  PIPSOLAR_TEXT_SENSOR(last_qpiri, QPIRI)
+  PIPSOLAR_TEXT_SENSOR(last_qmod, QMOD)
+  PIPSOLAR_TEXT_SENSOR(last_qflag, QFLAG)
+  PIPSOLAR_TEXT_SENSOR(last_qpiws, QPIWS)
+  PIPSOLAR_TEXT_SENSOR(last_qt, QT)
+  PIPSOLAR_TEXT_SENSOR(last_qmn, QMN)
+
+  PIPSOLAR_SWITCH(output_source_priority_utility_switch, QPIRI)
+  PIPSOLAR_SWITCH(output_source_priority_solar_switch, QPIRI)
+  PIPSOLAR_SWITCH(output_source_priority_battery_switch, QPIRI)
+  PIPSOLAR_SWITCH(input_voltage_range_switch, QPIRI)
+  PIPSOLAR_SWITCH(pv_ok_condition_for_parallel_switch, QPIRI)
+  PIPSOLAR_SWITCH(pv_power_balance_switch, QPIRI)
+
+  void switch_command(const std::string &command);
+  void setup() override;
+  void loop() override;
+  void dump_config() override;
+  void update() override;
+
+ protected:
+  static const size_t PIPSOLAR_READ_BUFFER_LENGTH = 110;  // maximum supported answer length
+  static const size_t COMMAND_QUEUE_LENGTH = 10;
+  static const size_t COMMAND_TIMEOUT = 5000;
+  uint32_t last_poll_ = 0;
+  void add_polling_command_(const char *command, ENUMPollingCommand polling_command);
+  void empty_uart_buffer_();
+  uint8_t check_incoming_crc_();
+  uint8_t check_incoming_length_(uint8_t length);
+  uint16_t cal_crc_half_(uint8_t *msg, uint8_t len);
+  uint8_t send_next_command_();
+  void send_next_poll_();
+  void queue_command_(const char *command, uint8_t length);
+  std::string command_queue_[COMMAND_QUEUE_LENGTH];
+  uint8_t command_queue_position_ = 0;
+  uint8_t read_buffer_[PIPSOLAR_READ_BUFFER_LENGTH];
+  size_t read_pos_{0};
+
+  uint32_t command_start_millis_ = 0;
+  uint8_t state_;
+  enum State {
+    STATE_IDLE = 0,
+    STATE_POLL = 1,
+    STATE_COMMAND = 2,
+    STATE_POLL_COMPLETE = 3,
+    STATE_COMMAND_COMPLETE = 4,
+    STATE_POLL_CHECKED = 5,
+    STATE_POLL_DECODED = 6,
+  };
+
+  uint8_t last_polling_command_ = 0;
+  PollingCommand used_polling_commands_[15];
```

Comment:
how about switching this to std::array and using the .at accessor? this way you won't have any over- or underruns